### PR TITLE
H 1848 balance not correct when delegating

### PR DESCRIPTION
**Pull request by RamRamez**
Issue #1848 and #1969

### CODE REVIEW

**Comment by aminlatifi** [comment]
@RamRamez You couldn't manage to fetch donation list after user opened the modal?

**Comment by aminlatifi** [comment]
https://github.com/Giveth/giveth-dapp/issues/2035

**Comment by RamRamez** [comment]
In this refactoring, the rerendering of the component is decreased from 4 to 2

**Comment by RamRamez** [comment]
> @RamRamez You couldn't manage to fetch donation list after user opened the modal?

Done

In this refactoring, the rerendering of the component is decreased from 4 to 2

File: src/components/DelegateMultipleButton.jsx [hunk 1]
```diff
@@ -97,9 +98,13 @@ const DelegateMultipleButton = props => {
 
   const dacsObserver = useRef(null);
 
-  const loadDonations = async ids => {
+  const loadDonations = async (ids, token) => {
```
Comment (aminlatifi) [inline src/components/DelegateMultipleButton.jsx:101]: It's not a good practice, loadDonations should be called on token change

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/components/DelegateMultipleButton.jsx [hunk 2]
```diff
@@ -198,75 +205,56 @@ const DelegateMultipleButton = props => {
...
-    }
-  }, [delegationOptions]);
-
-  useEffect(() => {
-    loadDonations(objectToDelegateFrom).then();
-  }, [selectedToken]);
+  const getDacs = () => {
+    const { milestone, campaign } = props;
+    const userAddress = currentUser ? currentUser.address : '';
+    dacsObserver.current = feathersClient
```
Comment (aminlatifi) [inline src/components/DelegateMultipleButton.jsx:211]: We don't need dacObserver anymore

File: src/components/DelegateMultipleButton.jsx [hunk 3]
```diff
@@ -198,75 +205,56 @@ const DelegateMultipleButton = props => {
...
+      });
+  };
 
   function openDialog() {
     isLoggedIn(currentUser)
       .then(() => checkBalance(balance))
-      .then(() => setModalVisible(true));
+      .then(() => {
+        setModalVisible(true);
+        getDacs();
```
Comment (aminlatifi) [inline src/components/DelegateMultipleButton.jsx:256]: In case I change the user, getDacs result is invalid

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/components/DelegateMultipleButton.jsx [hunk 4]
```diff
@@ -339,17 +328,157 @@ const DelegateMultipleButton = props => {
...
+  );
+
+  const modalLoading = (
+    <div>
+      <h2>Please wait while data is loading</h2>
+      <Loader />
+    </div>
+  );
+
+  const isContextReady = !whiteListIsLoading && !userContextIsLoading && Web3ContextIsEnabled;
```
Comment (aminlatifi) [inline src/components/DelegateMultipleButton.jsx:480]: I wish isContextReady not be true till the getDacs fetched and filled the options

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/components/DelegateMultipleButton.jsx (hunk 4 above)
Comment (RamRamez) [reply src/components/DelegateMultipleButton.jsx]: done

**Pull request merged**